doc/man1/mkcmclcon.1
.TH MKCMCLCON 1 "2010-12-23" 1.0 "LINX"
.SH NAME
mkcmclcon \- create LINX Connection Manager Control Layer connections.
.SH SYNOPSIS
.B mkcmclcon [OPTIONS] <connection>

.SH DESCRIPTION
Create a LINX Connection Manager Control Layer connection to a remote node.

.SH OPTIONS

.IP "-h, --help"
Display help and exit.

.IP "-c, --connection"
What connection to be used for this CMCL connection. The CMCL is commonly used
on top of another LINX Connection Manager. But, this other Connection Manager
can be very thin as the CMCL provides connection handling and supervision.

.IP "-s, --silent"
Silent mode, only error messages are printed to stdout.

.IP "-t, --con_tmo"
Connection timeout is used for supervision of the connection. The CMCL will
send three alive packets during this timeout. If it has not received an alive
packet from the peer upon sending the fourth alive packet, it will consider the
connection as disconnected. Default value is 1000 (ms).
If the CMTL layer has its own form of detection of a peer disconnection, it's
safe to turn off the keep-alive in CMCL.
Value of 0xFFFFFFFF should be used to turn off keep-alive in CMCL.

.SH FILES
None.

.SH DIAGNOSTICS
None.

.SH KNOWN BUGS
None.

.SH EXAMPLES
.nf
Create a CMCL connection between node A and node B. In this example, node A and
node B have created a LINX ethernet connection to each other (ethcm_A and 
ethcm_B). Since ECM layer has its own implementation of a state connection
machine, the CMCL keep-alive mechanism must be turned off.

On node A,
mkcmclcon -t 0xFFFFFFFF -c ethcm/ethcm_B node_B

On node B,
mkcmclcon -t 0xFFFFFFFF -c ethcm/ethcm_A node_A
.fi

.SH "SEE ALSO"
.BR linx "(7), "
.BR linxcfg "(1), "
.BR linxstat "(1), "
.BR linxdisc "(8), "
.BR mklink "(1), "
.BR mkethcon "(1), "
.BR rmcmclcon "(1)"

.SH AUTHOR
Enea LINX team

.SH COPYRIGHT
Copyright (c) 2006-2017, Enea Software AB
All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

Redistributions of source code must retain the above copyright notice, this
list of conditions and the following disclaimer.
Redistributions in binary form must reproduce the above copyright notice,
this list of conditions and the following disclaimer in the documentation
and/or other materials provided with the distribution.
Neither the name of Enea Software AB nor the names of its
contributors may be used to endorse or promote products derived from this
software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.


ChangeLog
Version 2.6.9
==========
* Fix warnings (LINXCR-2369)
* Print pool alloc kernel parameters in linxstat -p (LINXCR-2414)
* Fix TCP listen() backlog value to handle slow connections (LINXCR-2412)
* Fix legacy issue between LINX For Linux 2.6.9 kernel modules and liblinx from
  older releases (LINXCR-2413) 
* Fix related to host architecture identifying for powerpc (LINXCR-2410)
* Fix related to RX and TX tasklets not being correctly killed (LINXCR-2408)
* Fix race condition in ECM and RIOCM when link is destroyed (LINXCR-2407)
* Add workaround in ETH CM and RIO CM to protect against tasklet preempt_rt bug
* (https://bugzilla.kernel.org/show_bug.cgi?id=61451) (LINXCR-2406)
* Update makefiles and create autogen.sh to regenerate automake files (LINXCR-2363)
* Add LINX patches for libpcap-1.9.0 and tcpdump-4.9.2 (LINXCR-2365)
* Fix linxtracedrv for building on RHEL >= 7.5 (LINXCR-2339)
* Add support for Linux kernels up to 4.19 version (LINXCR-2362)
* Fix issues and enhance the pool feature in LINX For Linux (LINXCR-2358)
* Add kernel module parameter to specify the size of the linx_receive() buffer (LINXCR-2360)
* Fix recvfrom()/recvmsg() to work with linx_send() when in the same process (LINXCR-2290)
* Update documentation regarding the usage example of recvmsg() with LINX sockets (LINXCR-2286)
* Fix issue when trying to create multiple TCP CM IPv6 connections (LINXCR-2329)
* Automate Open Source Report and Copyright Scan generation (LINXCR-2231)
* Add option to allocate signals using linx_s_alloc in linx_bmark app (LINXCR-2244)
* Fix get_time function to handle time wrapping (LINXCR-2291)
* Fix linxcfg ETH CM connection with MTU (LINXCR-2274)
* Fix linxcfg TCP CM connection over IPv6 (LINXCR-2223)
* Documentation update: unmark linxcfg as being obsolete (LINXCR-2226)
* Add LINX patches for libpcap-1.8.1 and tcpdump-4.9.1 (LINXCR-2225)

Version 2.6.8 (2017-06-09)
==========
* Fix bug related to struct msghdr initialization in TCP CM (LINXCR-2200)
* Fix CentOS 7.3 build (LINXCR-2200)
* Fix sending signals of maximum allowed size in LINX Gateway (LINXCR-1985)
* Fix LINX pool allocation at specific physical address (LINXCR-1979)
* Fix double free from LINX pool on linx_send_w_s()/linx_send_w_opt() (LINXCR-2203)
* Fix warn_slowpath kernel warning, reported with ERRORCHECKS_MEM=yes (LINXCR-1977)
* Remove "octeon-001.patch" application on "config.sub". (LINXCR-2056)
* Fix an issue in linxgws related to a gcc bug. (LINXCR-2044)
* Add support in linx_open() for shared virtual pools (LINXCR-1903)
* Add support for CentOS 7.2 kernel 3.10.0-327.4.4.el7.x86_64 (LINXCR-1997)
* Add support for up to 4.4 kernel version. (LINXCR-2015)
* Add SOCK_CLOEXEC option at linx socket creation. (LINXCR-1995)
* Fix kernel module backwards compatibility issue in af_linx.c. (LINXCR-2051)
* Fix linx_send() to be in synch with the manual (LINXSC-212, LINXCR-1805)
* Add support to disable LINX signal pool feature (LINXCR-2196)

Version 2.6.7 (2015-11-06)
==========
* Support for 3.16, 3.17, 3.18 and 4.1 kernel versions.
* Fix debug options ERRORCHECKS and ERRORCHECKS_MEM. (LINXCR-1927)
* Fix buffer overflow issue in libgw. (LINXCR-1934)
* Fix gcc related issue in linxdisc. (LINXCR-1933)
* Fix issue when creating a link over two connections, one on a physical
  network interface, and the other on vlan interface over the same
  physical interface. (LINXCR-1935)
* Fix issue in ECM that was causing garbage data to be received. (LINXCR-1975)
* Fix pool issue when memory is freed. (LINXCR-1778)
* Fix some build warnings. (LINXCR-1965)
* Fix buffer overflow issue in linxcfg. (LINXCR-1967)
* Fix buffer space problem for linx_hunt (LINXSC-270, LINXCR-1926)
* Fix kernel crash when creating connection over alias interface (LINXCR-1850)
* Fix some race conditions in ecm that prevents
RLNH to complete IOCTL delete(ELEMSC-1545, LINXCR-1896)
* Fix some severity 1 and 2 issues. (LINXSC-244, LINXCR-1894)
* Linearize received SKB in ecm_rx interrupt.
* Renamed libcfg.a to liblinxcfg.a

Version 2.6.6 (2014-03-24)
==========
* Fix memory leak in linx_send.
* Remove several warnings when building linx_bmark, linxgw, linxcfg.
* LINX patches for libpcap 1.3.0 and tcpdump 4.3.0 added.
* Add missing include file stdlib.h to print-linx.c in the tcpdump patch
  for linxdump.
* linxstat prints all error messages to stderr.
* linxdisc will never timeout a connection as long as the link is up.
* Support for 3.11 kernel versions.
* Fix a crash in the Ethernet CM when using bonded interfaces.
* Fix NULL pointer dereference in ecm_conn.c

Version 2.6.5 (2013-05-30)
==========
* Support for 3.8 kernel versions.
* Fix a bug in linx_ioctl_pool_info that caused the pool size to be read
  as 0 on powerpc 64 bit big endian machines.

Version 2.6.4 (2012-09-13)
==========
* Support for 3.2 kernel versions.
* linxstat fix for pending hunts with long huntnames.
* Change heap memory allocation in db_proc_write to fix Linux crash.
* Enable LINX to work in context of network namespaces.

Version 2.6.3 (2012-01-18)
==========
* Fix: CMCL-in case heartbeat mechanism is disabled, send CONN_ALV message
in response for remote side keep-alive message.
* LINX support for Power Save modes.

Version 2.6.2 (2011-12-16)
==========
* allow disabling heartbeat mechanism when creating ETH-CM connections.
* allow disabling heartbeat mechanism when creating CMCL connections.
* fix rio-cm connection mechanism.
* fix skb allocations in rio-cm (TX side).
* fix hunt/attach/request_tmo ioctl handlers for 32-bit machines.

Version 2.6.1 (2011-11-10)
==========
* fix pool locking mechanism.

Version 2.6.0 (2011-10-14)
==========
* fix pool signals END_MARK setting for buffers allocated by CMs.
* fix pool buffer true_size value in sig_adm section, to allow correct
  signal resizing.
* fix pool buffer owner setting in get_mem.
* fix building errors for kernel versions < 2.6.22.
* linx_recvmsg now accepts pool signals. linx_info_* handles pool signals
  correctly.

linxcfg/mkcmclcon.c
/*
 * Copyright (c) 2006-2010, Enea Software AB
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * Redistributions of source code must retain the above copyright notice, this
 * list of conditions and the following disclaimer.
 * Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * Neither the name of Enea Software AB nor the names of its
 * contributors may be used to endorse or promote products derived from this
 * software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
#include <errno.h>
#include <getopt.h>
#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <string.h>
#include <linxcfg.h>

#include "db_utils.h"
#include "cmcl_utils.h"

#define STR(s) #s
#define STRSTR(s) STR(s)

#define DEF_CON_TMO 300

struct option long_opts[] = {
        {"help", 0, NULL, (int)'h'},
        {"silent", 0, NULL, (int)'s'},
        {"con_tmo", 1, NULL, (int)'t'},
        {"connection", 1, NULL, (int)'c'},
        {NULL, 0, NULL, 0}
};

static char *usage =
"mkcmclcon [OPTIONS] <connection>\n"
"Create a CMCL connection.\n"
"OPTIONS:\n"
"    -h, --help\n"
"        Print this text.\n"
"    -c, --connection\n"
"        connection to use for the cmcl connection.\n"
"    -s, --silent\n"
"        Only error messages are printed to stdout.\n"
"    -t, --con_tmo\n"
"        Time-out value in milli-sec for the alive packet.\n"
"        Default is " STRSTR(DEF_CON_TMO) " millisec.\n"
"        Value of 0xFFFFFFFF should be used to turn off keep-alive.\n"
"Example:\n"
"    mkcmclcon -c ethcm/conn_b node_b\n";

static uint32_t get_u32(const char *opt, const char *arg)
{        
        char *end;
        uint32_t u32;

        u32 = (uint32_t)strtoul(arg, &end, 0);
        if (*end != '\0') {
                printf("mkshmcon: error in %s\n"
                       "Try 'mkshmcon --help' for more information\n",
                       opt);
                exit(1);
        }
        return u32;
}

int main(int argc, char *argv[])
{
        int c, status, silent;
        char *s;
        union linx_con_arg con;

        memset(&con.cmcl, 0, sizeof(con.cmcl));
        con.cmcl.con_tmo = DEF_CON_TMO; 
        silent = 0;

        for (c = 0; c != -1;) {
                c = getopt_long(argc, argv, "hst:c:", long_opts, NULL);
                switch (c) {
                case 'h':
                        printf("Usage: %s\n", usage);
                        return 0;
		case 'c':
			con.cmcl.con_name = optarg;
			break;
                case 's':
                        silent = 1;
                        break;
                case 't':
                        con.cmcl.con_tmo = get_u32("--con_tmo", optarg);
                        break;
                case '?':
                        return 1;
                }
        }
        if (argv[optind] == NULL) {
                printf("mkcmclcon: connection name missing\n");
                return 1;
        }

        con.cmcl.name = argv[optind];

        s = NULL;
        status = linx_create_connection(LINX_CON_CMCL, &con, &s);
        if (status != 0) {
                printf("mkcmclcon: couldn't create connection '%s': %s\n",
                       con.cmcl.name, strerror(status));
                return 1;
        }
        
        if (!silent)
                printf("mkcmclcon: created connection '%s'.\n"
                       "Now use 'mklink -c %s ...' to create a link.\n",
                       s, s);
        free(s);
        return 0;
}
